core/corehttp/gateway_handler.go

```diff
@@ -146,15 +154,14 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		return
 	}
 
-	// set these headers _after_ the error, for we may just not have it
-	// and dont want the client to cache a 500 response...
-	// and only if it's /ipfs!
-	// TODO: break this out when we split /ipfs /ipns routes.
+	// Remember to unset the Cache-Control/ETag headers before error
+	// responses!  The webError functions unset them automatically.
+	// TODO: Consider letting clients cache 404s.  Are there cases where
+	// that would hurt?
```

Comment:
It’s unclear to me where it would hurt, though. Given a successfully resolved `/ipfs/QmFoo`, the knowledge that “`/ipfs/QmFoo/fileA` exists” should be as cacheable as “`/ipfs/QmFoo/fileB` does not exist” for example. `QmFoo` is immutable after all.

The same applies to `/ipns/QmBar`. The record’s TTL dictates that the knowledge that “`/ipns/QmBar/fileA` exists” can be cached for just as long as that “`/ipns/QmBar/fileB` does not exist”.